sdk/servicebus/service-bus/src/core/batchingReceiver.ts

```diff
@@ -325,200 +329,202 @@ export class BatchingReceiverLite {
     const brokeredMessages: ServiceBusMessageImpl[] = [];
     const loggingPrefix = `[${receiver.connection.id}|r:${receiver.name}]`;
 
-    return new Promise<ServiceBusMessageImpl[]>((resolve, reject) => {
-      let totalWaitTimer: NodeJS.Timer | undefined;
+    let totalWaitTimer: NodeJS.Timer | undefined;
 
-      // eslint-disable-next-line prefer-const
-      let cleanupBeforeResolveOrReject: () => void;
+    // eslint-disable-next-line prefer-const
+    let cleanupBeforeResolveOrReject: () => void;
 
-      const onError: OnAmqpEvent = (context: EventContext) => {
-        cleanupBeforeResolveOrReject();
+    const reject = (err: Error | AmqpError) => {
+      cleanupBeforeResolveOrReject();
+      origReject(err);
+    };
 
-        const eventType = context.session?.error != null ? "session_error" : "receiver_error";
-        let error = context.session?.error || context.receiver?.error;
-
-        if (error) {
-          error = translateServiceBusError(error);
-          logger.logError(
-            error,
-            `${loggingPrefix} '${eventType}' event occurred. Received an error`
-          );
-        } else {
-          error = new ServiceBusError(
-            "An error occurred while receiving messages.",
-            "GeneralError"
-          );
-        }
-        reject(error);
-      };
+    const resolveImmediately = (result: ServiceBusMessageImpl[]) => {
+      cleanupBeforeResolveOrReject();
+      origResolve(result);
+    };
 
-      this._closeHandler = (error?: AmqpError | Error): void => {
+    const resolveAfterPendingMessageCallbacks = (result: ServiceBusMessageImpl[]) => {
+      // NOTE: through rhea-promise most of our event handlers are made asynchronous by calling setTimeout(emit).
```

Comment:
Thanks for adding this note.

Reply:
Yeah, it's nice to have an explanation for the weird edge cases. It's also nice to know when we can remove it. :)